synapse/storage/registration.py

```diff
@@ -598,11 +598,64 @@ def __init__(self, db_conn, hs):
             "user_threepids_grandfather", self._bg_user_threepids_grandfather,
         )
 
+        self.register_background_update_handler(
+            "users_set_deactivated_flag", self._set_deactivated_flag,
+        )
+
         # Create a background job for culling expired 3PID validity tokens
         hs.get_clock().looping_call(
             self.cull_expired_threepid_validation_tokens, THIRTY_MINUTES_IN_MS,
         )
 
+    @defer.inlineCallbacks
+    def _set_deactivated_flag(self, progress, batch_size):
+        last_user = progress.get("user_id", "")
+
+        def _set_deactivated_flag_txn(txn):
+            txn.execute(
+                """
+                SELECT
+                    users.name
+                FROM users
+                    LEFT JOIN access_tokens ON (access_tokens.user_id = users.name)
+                    LEFT JOIN user_threepids ON (user_threepids.user_id = users.name)
+                WHERE password_hash IS NULL OR password_hash = ''
+                AND users.name > ?
+                GROUP BY users.name
+                HAVING count(access_tokens.token) = 0
+                AND count(user_threepids.address) = 0
```

Comment:
Ah, sorry, I think the needs to happen outside the SQL, otherwise its going to potentially pull a lot of data. E.g. if there aren't any deactivated accounts then this will load the entire users table all at once, which will be a bit slow.

You'll also want to add a `ORDER BY users.name ASC` clause, as we need the ordering to be consistent.